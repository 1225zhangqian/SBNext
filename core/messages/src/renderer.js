import logger, { colors } from '@sb/core-logger/node';
import * as documentation from './documentation';
import cleanIndent from '@sb/core-logger/clean-indent';

export const unloadable = ({ name }) => {
  const packageref = colors.blue(`@sb/renderer-${name}`);
  logger.error(
    cleanIndent(`
      ${colors.red.bold('Loading of a renderer failed.')}
      Please check your storybook config; you're using something like:
      
      ${colors.gray(`renderer: ['${colors.orange(name)}']`)},

      This maps to a npm package: ${packageref}

      Perhaps you need to install this package: 
      > ${colors.gray(`npm install ${packageref} --save`)}

      If it won't install perhaps you made a spelling mistake? Find the correct package name by searching npm:
      > ${colors.gray(`npm search @sb/renderer-${name}`)}
              
      Learn more about renderers at ${documentation.link(
        '/renderers'
      )}, this also includes a list of all renderers. 
      Alternatively you can search npm from your terminal:
      > ${colors.gray('npm search @sb/renderer')}
    `)
  );
};
export const unknown = ({ obj }) => {
  logger.error(
    cleanIndent(`
      ${colors.red.bold('The object supplied does not match the interface of a renderer.')}
      Please check the storybook config; you're using something like:
      
      ${colors.gray(`renderer: [${colors.orange(JSON.stringify(obj))}],`)}

      The object should match this interface:
      ${colors.gray(
        `{ name: ${colors.pink('string')}, dependencies ${colors.pink('Array<string>')} }`
      )}

      Learn more about renderers at ${documentation.link('/renderers')}
    `)
  );
};
export const foundInAsset = ({ name, assets }) => {
  logger.warn(
    cleanIndent(`
      The output/assets: ${assets
        .map(a => colors.orange(a))
        .join(', ')} appear to include a sb renderer!

      This could be a mistake. It won't likely break anything, but it very un-optimized.

      You should probably configure storybook to extract the renderer!

      You will likely need to add the renderer:
      ${colors.gray(`renderer: ['@sb/renderer-${colors.orange(name)}']`)},

      Learn more about configuration at ${documentation.link('/configuration')}
      Learn more about renderers at ${documentation.link('/renderers')}
    `)
  );
};
