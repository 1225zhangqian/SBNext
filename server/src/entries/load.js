/* global __NEXT_DATA__ */
import glob from 'glob';
import { sep } from 'path';
import fetch from 'unfetch';

import { isServer } from './env';
import processEntries from './process';

export default async (path = 'docs') => (isServer() ? fromServer(path) : fromClient(path));

const fromServer = async entriesPath => {
  const paths = glob.sync(`${entriesPath}/**/*.md`, { root: process.cwd() });

  return processEntries(paths, entriesPath).filter(({ data }) => data.published !== false);
};

const fromClient = async path => {
  // in safari the popstate event is fired when user press back and
  // causes the getInitialProps to be called in the SSR version
  // This will pickup the current props and return it as a workaround
  // https://github.com/zeit/next.js/issues/2360
  if (__NEXT_DATA__.nextExport) {
    return __NEXT_DATA__.props.docs;
  }
  const resp = await fetch('/_load_entries');
  return resp.json();
};

export const byFileName = async (path, root = 'docs') =>
  isServer() ? byFileNameFromServer(path, root) : byFileNameFromClient(path);

const byFileNameFromServer = (path, entriesPath) => processEntries([path], entriesPath).pop();

const byFileNameFromClient = async path => {
  // this is used to make next.js Link to work on exported sites.
  if (__NEXT_DATA__.nextExport) {
    return findDocFromNextCache(path);
  }
  const resp = await fetch(`/_load_entry/${path.replace(sep, '/')}`);
  return resp.json();
};

const findDocFromNextCache = path => {
  const { doc, docs } = __NEXT_DATA__.props;

  return doc && doc.data._entry === path
    ? doc
    : docs.filter(p => p.data._entry === path).reduce(v => v);
};
