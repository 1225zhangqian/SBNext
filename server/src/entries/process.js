import { readFileSync, statSync } from 'fs';
import { resolve, basename, extname, relative, dirname, sep } from 'path';
import fm from 'frontmatter';
import pathToRegEx from 'path-to-regexp';

export default paths =>
  paths
    .map(path => readFileSync(path, 'utf-8'))
    .map(fm)
    .map(addEntry(paths))
    .map(addName)
    .map(addDate)
    .map(addUrl);

const addPage = value => {
  const { data } = value;
  const { page = 'doc' } = data;
  return { ...value, data: { ...data, page } };
};

const addEntry = paths => (value, idx) => {
  const { data } = value;
  return { ...value, data: { ...data, _entry: paths[idx] } };
};

const addName = value => {
  const { data } = value;
  return { ...value, data: { ...data, name: createName({ ...data }) } };
};

const addUrl = value => {
  const { data } = value;
  return { ...value, data: { ...data, url: createURL({ ...data }) } };
};

const addDate = value => {
  const { data } = value;
  return { ...value, data: { ...data, date: createDate({ ...data }) } };
};

const DATE_IN_FILE_REGEX = /^(\d{4}-\d{2}-\d{2})-(.+)$/;
const DATE_MATCH_INDEX = 1;
const NAME_MATCH_INDEX = 2;

const extractFileName = path => basename(path, extname(path));

const createName = ({ _entry }) => {
  const name = extractFileName(_entry);
  const match = name.match(DATE_IN_FILE_REGEX);

  return match ? match[NAME_MATCH_INDEX] : name;
};

const createURL = data => {
  const { _entry } = data;

  return `/${_entry.replace(/\.[^/.]+$/, '/')}`;
};

const createDate = ({ _entry, date }) => {
  const name = extractFileName(_entry);
  const match = name.match(DATE_IN_FILE_REGEX);
  let result;

  if (date) {
    result = new Date(date);
  } else if (match) {
    result = new Date(match[DATE_MATCH_INDEX]);
  } else {
    result = fileCreationDate(_entry);
  }

  return result.toJSON();
};

const fileCreationDate = path => {
  const { birthtime } = statSync(path);
  return birthtime;
};
