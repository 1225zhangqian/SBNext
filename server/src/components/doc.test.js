import React from 'react';
import renderer from 'react-test-renderer';

import unified from 'unified';
import { byFileName } from '../entries/load';
import withDoc, { Content } from '../components/doc';

jest.mock('../entries/load');
jest.mock('unified', () => {
  class Impl {
    use = mockedUnified.use;
    processSync = mockedUnified.processSync;
  }
  function mockedUnified() {
    return new Impl();
  }
  mockedUnified.processSync = jest.fn();
  mockedUnified.use = jest.fn(function mockUse() {
    return this;
  });

  return mockedUnified;
});

describe('withDoc', () => {
  test('exports HOC withDoc as default', () => {
    expect(withDoc).toBeDefined();
  });

  test('withDocs should add `doc` property to getInitialProps', async () => {
    const expected = { data: {}, content: '' };
    const expectedFileName = 'fake';
    const Component = withDoc(() => <div>Test</div>);

    byFileName.mockReturnValueOnce(expected);
    const actual = await Component.getInitialProps({ query: { _entry: expectedFileName } });

    expect(actual.doc).toBeDefined();
    expect(actual.doc).toEqual(expect.objectContaining(expected));
    expect(byFileName).toHaveBeenCalledWith(expectedFileName);
  });
});

describe('Content', () => {
  test('exports Content', () => {
    expect(Content).toBeDefined();
  });

  test('Content component should render doc content', () => {
    const expectedContent = 'lorem ipsum';

    unified.processSync.mockReturnValueOnce({ contents: <p>{expectedContent}</p> });

    const comp = renderer.create(<Content content={expectedContent} />);

    expect(unified.processSync).toHaveBeenCalledWith(expectedContent);
    expect(comp.toJSON()).toMatchSnapshot();
  });
});
