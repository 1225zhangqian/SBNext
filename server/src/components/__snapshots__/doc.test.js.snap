// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Content Content component should render doc content 1`] = `
<p>
  lorem ipsum
</p>
`;
